server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java

```diff
@@ -215,20 +215,22 @@ private void connectToNodeOrRetry(
             ActionListener.wrap(
                 conn -> connectionValidator.validate(conn, resolvedProfile, ActionListener.runAfter(ActionListener.wrap(ignored -> {
                     assert Transports.assertNotTransportThread("connection validator success");
+                    final var managerRefs = AbstractRefCounted.of(conn::onRemoved);
```

Comment:
I tried following the code and the PR description to understand the edge case. I see finally that it's related to this part of the code a bit above in this file:

```
            final int failureCount = previousFailureCount + 1;
            if (failureCount < 10) {
                logger.trace("concurrent connect/disconnect for [{}] ([{}] failures), will try again", node, failureCount);
                connection.addRemovedListener(
                    delegate.delegateFailure(
                        (retryDelegate, ignored) -> connectToNodeOrRetry(
                            node,
                            connectionProfile,
                            connectionValidator,
                            failureCount,
                            retryDelegate
                        )
                    )
                );
```

So basically the connection could have been closed, and the onRemoved called before the `finally` below removes it from the pendingConnections. And so it would try connecting again to it. I see!

Reply:
That sounds right, but I don't think it is especially related to the code you quoted. This bug can happen even without this retrying.